lib/ukstore/store.c

```diff
@@ -29,3 +29,1267 @@
+		__u64 val;
+
+		str = calloc(_U64_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U64_STRLEN, "%" __PRIu64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s64): {
+		__s64 val;
+
+		str = calloc(_S64_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S64_STRLEN, "%" __PRIs64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(uptr): {
+		__uptr val;
+
+		str = calloc(_UPTR_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.uptr(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _UPTR_STRLEN, "0x%" __PRIx64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(charp): {
+		char *val = NULL;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		*out  = val;
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+
+int
+_uk_store_get_ncharp(const struct uk_store_entry *e, char **out, size_t maxlen)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	case UK_STORE_ENTRY_TYPE(u8): {
+		__u8 val;
+
+		ret = e->get.u8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "%" __PRIu8, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s8): {
+		__s8 val;
+
+		ret = e->get.s8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "%" __PRIs8, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u16): {
+		__u16 val;
+
+		ret = e->get.u16(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "%" __PRIu16, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s16): {
+		__s16 val;
+
+		ret = e->get.s16(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "%" __PRIs16, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u32): {
+		__u32 val;
+
+		ret = e->get.u32(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "%" __PRIu32, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s32): {
+		__s32 val;
+
+		ret = e->get.s32(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "%" __PRIs32, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u64): {
+		__u64 val;
+
+		ret = e->get.u64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "%" __PRIu64, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s64): {
+		__s64 val;
+
+		ret = e->get.s64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "%" __PRIs64, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(uptr): {
+		__uptr val;
+
+		ret = e->get.uptr(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(*out, maxlen, "0x%" __PRIx64, val);
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(charp): {
+
+		ret = e->get.ncharp(e->cookie, out, maxlen);
```

Comment:
Hum :thinking:. The entry is type of `charp` but a `ncharp` callback is called? I was assuming that we take the `charp` interface, do an `snprintf()` and then free the allocated buffer before returning this function. Of course, this is not zero-copy anymore but hopefully the `charp` interface is most rarely used.

Reply:
makes sense, especially if we are removing `get.ncharp` anyway